zypp/RepoStatus.cc

```diff
@@ -12,16 +12,41 @@
 #include <iostream>
 #include <sstream>
 #include <fstream>
+#include <set>
 #include <zypp/base/Logger.h>
 #include <zypp/base/String.h>
 #include <zypp/RepoStatus.h>
+#include <zypp/RepoInfo.h>
 #include <zypp/PathInfo.h>
 
 using std::endl;
 
 ///////////////////////////////////////////////////////////////////
 namespace zypp
-{ /////////////////////////////////////////////////////////////////
+{
+  ///////////////////////////////////////////////////////////////////
+  namespace
+  {
+    /** Recursive computation of max dir timestamp. */
+    void recursiveTimestamp( const Pathname & dir_r, time_t & max_r )
+    {
+      std::list<std::string> dircontent;
+      if ( filesystem::readdir( dircontent, dir_r, false/*no dots*/ ) != 0 )
+	return; // readdir logged the error
+
+      for_( it, dircontent.begin(), dircontent.end() )
+      {
+	PathInfo pi( dir_r + *it, PathInfo::LSTAT );
+	if ( pi.isDir() )
+	{
+	  if ( pi.mtime() > max_r )
+	    max_r = pi.mtime();
+	  recursiveTimestamp( pi.path(), max_r );
+	}
+      }
+    }
```

Comment:
Or for the implementation of `filesystem::readdir`. As long as we have SLE12 in maintenance, I'd prefer the `filesystem::` functions in all code that might become back ported.

Reply:
Right, however the std API does not require a extra PathInfo() object, it has all required informations in the object returned by the iterator. I'd want to avoid calling stat() twice for every directory entry.

Reply:
But where do we call it twice? We do one `::readdir` and one `::stat` in each `PathInfo` we create. I don't think `filesystem::readdir` can do fewer stats.

Reply:
We don't call it twice atm. But if we'd use the std::recursive_directory_iterator to implement our filesystem::readdir() it would already call stat() in there and also again when we create the PathInfo(). We'd need to check when the std implementation calls it ( might be lazy ) before we reimplement it

Reply:
Now I got your point. I can't imagine that they do not do it lazy... but better check it.